WebGPURenderer: Fixed shadows not rendering correctly with `logarithmicDepthBuffer`

PoseidonEnergy opened the pull request:
**Update 2024-11-13 | This post and info on the logarithmic depth buffer calculation is outdated. Please see the new revisions to the logarithmic depth functions in the pull request here: https://github.com/mrdoob/three.js/pull/29870**

####################################################################
####################################################################
####################################################################

Related issue: Closes #29200 

This pull request corrects the calculations for the logarithmic depth buffer, and ensures proper depth value adjustment for shadowmaps when logarithmicDepthBuffer = true and shadows are enabled.
I found 4 bugs that are causing the issues with the logarithmic depth buffer; I was able to fix 3 of them and write a workaround for 1 of them:

### Bug 1
**Location** | `NodeMaterial.setupDepth()` in **src/materials/nodes/NodeMaterial.js**
**Details** | In the following line, the variable `fragDepth` always equals "2" when a shadow is being rendered, no matter what the actual depth of the frag is (i.e. the `w` component of `modelViewProjection()` always equals "1"):

`const fragDepth = modelViewProjection().w.add(1);`

**Fix** | The line was replaced with the following:

`const fragDepth = viewZToOrthographicDepth(positionView.z, cameraNear, cameraFar)`

**Screenshot** | (see bug no. 2)

### Bug 2
**Location** | `NodeMaterial.setupDepth()` in **src/materials/nodes/NodeMaterial.js**
**Details** | In the following line, the logarithmic depth calculation is not correct:

`depthNode = fragDepth.log2().mul(cameraLogDepth).mul(0.5);`

**Fix** | The following lines were added to **src/nodes/display/ViewportDepthNode.js** (note the change to the logarithmic depth calculation):

```js
export const perspectiveDepthToOrthographicDepth = ( depth, near, far ) => viewZToOrthographicDepth( perspectiveDepthToViewZ( depth, near, far ), near, far );`
export const orthographicDepthToLogarithmicDepth = ( depth, logDepthBufFC ) => depth.log2().mul( logDepthBufFC ).mul( 0.25 ).add( 1 );
```
And the former `depthNode` calculation line was replaced by a call to the new `orthographicDepthToLogarithmicDepth` function:

`depthNode = orthographicDepthToLogarithmicDepth( fragDepth, cameraLogDepth );`

**Screenshot** | *webgpu_camera_logarithmicdepthbuffer.html* results before and after change:

![image](https://github.com/user-attachments/assets/98140564-e952-414d-a75a-321dafa91357)

### Bugs 3a & 3b
**Location** | `AnalyticLightNode.setupShadow()` in **src/nodes/lighting/AnalyticLightNode.js**
**Details** | **3a)** The necessary adjustment to `coordZ` is not made when `logarithmicDepthBuffer` is `true`. **3b)** The aforementioned adjustment of `coordZ` requires use of the `cameraNear`, `cameraFar`, and `cameraLogDepth` uniforms, but these uniforms do not appear to be using the `shadow.camera` object for their calculations.
**Fix** | The following lines were added so that the shadow `coordZ` is adjusted for the logarithmic depth buffer, with the necessary uniforms correctly using the shadow camera object:

```js
if ( renderer.logarithmicDepthBuffer === true ) {

    // the normally available cameraNear, cameraFar, and cameraLogDepth nodes cannot be used here because they do
    // not get updated to use the shadow camera, so we have to declare our own "local" nodes here.
    // TODO: can we fix the cameraNear/cameraFar/cameraLogDepth nodes in src/nodes/accessors/Camera.js so we don't have to declare local ones here?
    if ( shadow.camera.isPerspective ) {

        const cameraNearLocal = uniform( 'float' ).onRenderUpdate( () => shadow.camera.near );
        const cameraFarLocal = uniform( 'float' ).onRenderUpdate( () => shadow.camera.far );
        coordZ = perspectiveDepthToOrthographicDepth( coordZ, cameraNearLocal, cameraFarLocal );

    }

    const cameraLogDepthLocal = uniform( 'float' ).onRenderUpdate( () => 2.0 / ( Math.log( shadow.camera.far + 1.0 ) / Math.LN2 ) );
    coordZ = orthographicDepthToLogarithmicDepth( cameraLogDepthLocal );

}
```

**Screenshot** | image is from a JSFiddle found at https://github.com/mrdoob/three.js/issues/29200

Note: the "shadow acne" in the "After" image is a GOOD thing to see, since it means the logarithmic depth shadowmap comparison is working as intended. The shadow acne is easily removed by setting `shadow.bias` to -0.0005 or some number close to that.

![image](https://github.com/user-attachments/assets/bb68ad2f-0e4d-4c20-9c95-c7d48c8d7373)

PoseidonEnergy committed:
Fixed shadows not rendering correctly when logarithmicDepthBuffer = true (issue #29200)

WestLangley commented:
I would not expect orthographic shadow cameras to be using a logarithmic depth buffer.

I'd suggest having a look at this line:

https://github.com/mrdoob/three.js/blob/d4164b8804c56f716935ce4af63ad4ddbf56e0a3/src/materials/nodes/NodeMaterial.js#L236

Compare with the WebGLRenderer implementation:

https://github.com/mrdoob/three.js/blob/d4164b8804c56f716935ce4af63ad4ddbf56e0a3/src/renderers/shaders/ShaderChunk/logdepthbuf_fragment.glsl.js#L6

/ping @sunag

PoseidonEnergy commented:
> I would not expect orthographic shadow cameras to be using a logarithmic depth buffer.
> 
> I'd suggest having a look at this line:
> 
> `depthNode = fragDepth.log2().mul( cameraLogDepth ).mul( 0.5 ); `
> 
> Compare with the WebGLRenderer implementation:
> 
> `gl_FragDepth = vIsPerspective == 0.0 ? gl_FragCoord.z : log2( vFragDepth ) * logDepthBufFC * 0.5;`

I looked into that as well. However while researching I saw this statement by @gkjohnson when that particular check was put into the shader:

from https://github.com/mrdoob/three.js/pull/17442#issue-490581179:
> This disables the logarithmic depth buffer when using an orthographic projection matrix so there are no benefits to using a logarithmic depth buffer with an ortho camera. It just changes the shader code to produce sensible results.

I interpreted that as: "theoretically there are benefits to using a logarithmic depth with an orthographic camera, but I had to disable it to fix this bug".

Or was I incorrect, and there are never benefits to using a logarithmic depth with an orthographic camera?

PoseidonEnergy committed:
Further improvements to the logarithmic depth buffer shadows bugfix:
1) Disabled logarithmic depth buffer for orthographic cameras.
2) Removed "cameraLogDepth" node from Camera.js as it is no longer used in the perspective-to-logarithmic depth conversion.
3) Ensured that modelViewProjection.w is used for perspective-to-logarithmic depth conversion, and that positionView.z is always used for orthographic cameras.
4) Adapted Outerra's logarithmic depth buffer formula, re-added "C" constant for increased resolution close to the camera, per Outerra's findings.
(issue #29200, PR #29447)

sunag commented on src/materials/nodes/NodeMaterial.js:
```diff
@@ -231,9 +231,11 @@ class NodeMaterial extends Material {
+				// to also include it in the shadow depth correction in AnalyticLightNode.setupShadow()!
+				const logarithmicDepth = perspectiveDepthToLogarithmicDepth( modelViewProjection().w, cameraFar );
+				const orthographicDepth = viewZToOrthographicDepth( positionView.z, cameraNear, cameraFar );
+				depthNode = cameraIsPerspective.equal( 1 ).select( logarithmicDepth, orthographicDepth );
```
Try using `builder.camera.isPerspectiveCamera` instead of `camera Perspective`. I think we can remove `cameraIs*` nodes.

sunag commented on src/nodes/lighting/AnalyticLightNode.js:
```diff
@@ -311,20 +312,34 @@ class AnalyticLightNode extends LightingNode {
+				shadowCoord = shadowCoord.xyz.div( shadowCoord.w );
+				if ( renderer.coordinateSystem === WebGPUCoordinateSystem ) {
+
+					coordZ = shadowCoord.z.mul( 2 ).sub( 1 ); // WebGPU: Conversion [ 0, 1 ] to [ - 1, 1 ]
```
We need a version for coordZ for WebGLBackend as well, this is crashing in E2E tests.

PoseidonEnergy committed:
Further improvements to the logarithmic depth buffer shadows bugfix:
1) Removed "cameraLogDepth" node from Camera.js as it is no longer used in the perspective-to-logarithmic depth conversion.
2) Removed "cameraIsPerspective" and "cameraIsOrthographic" nodes that were added in a previous commit
3) Ensured that modelViewProjection.w is used for perspective-to-logarithmic depth conversion, and that positionView.z is always used for orthographic cameras.
4) Adapted a modified version of Outerra's logarithmic depth buffer without requiring a "C" constant
(issue #29200, PR #29447)

PoseidonEnergy committed:
Further improvements to the logarithmic depth buffer shadows bugfix:
1) fixed bug where the shadowmap's "coordZ" was not getting set when not using WebGPU
2) fixed typos in comments
(issue #29200, PR #29447)

PoseidonEnergy commented:
@sunag I committed some updates that resolved the issues you pointed out:

1. The depth calculation uses `builder.camera.isPerspectiveCamera` now instead of `cameraIsPerspective`, and I removed the `cameraIsPerspective` and `cameraIsOrthographic` nodes.
2. I ensured the `coordZ` node is getting updated when using non-WebGPU coordinate system (this resolved the E2E tests).
3. I modified the logarithmic depth calculation to incorporate both the near plane and far plane (previously it only incorporated the far plane).

WestLangley commented:
@PoseidonEnergy Thank you for all your efforts so far.

@Mugen87 Can you please provide feedback to the OP? 🙏 I am concerned the OP is working without the benefit of guidance.

sunag committed:
Merge branch 'dev' into pr/29447

sunag committed:
cleanup

Spiri0 commented:
I noticed that the logarithmicDepthBuffer calculation has changed and now it no longer works for me.
I have a small codePen that I use in the forum to demonstrate logarithmicDepth with wgsl code.
@PoseidonEnergy If there were any errors, I'm happy that they were fixed. Fortunately, I never faced the errors. Can you help me adapt the codePen so that it works correctly again with r170?
 https://codepen.io/Spiri0/pen/vYqpRga?editors=0010

PoseidonEnergy commented:
> I noticed that the logarithmicDepthBuffer calculation has changed and now it no longer works for me. I have a small codePen that I use in the forum to demonstrate logarithmicDepth with wgsl code. @PoseidonEnergy If there were any errors, I'm happy that they were fixed. Fortunately, I never faced the errors. Can you help me adapt the codePen so that it works correctly again with r170? https://codepen.io/Spiri0/pen/vYqpRga?editors=0010

@Spiri0 Sure, happy to help--for reference, the latest logarithmic-depth-related pull request can be found here: https://github.com/mrdoob/three.js/pull/29561

In that latest pull request, the logarithmic depth formula was refined to following:

```
cameraNear = max( 1e-6, cameraNear ); // clamp cameraNear so we don't divide by 0
var logDepth = log2( ( w / cameraNear ) + 1 ) / log2( ( camerafar / cameraNear ) + 1 );
```

Regarding your code, the property `outPosition.w` can directly be used for `w` in the formula above (do not add `1` to it).
Additionally, the old uniform `logDepthBufFC` was completely removed. Lastly, the calculation now uses `camera.near` to fix small visual artifacts on objects extemely close to the camera (rare but definitely visible).

Here is a Desmos link if you want to visualize the new logarithmic depth curve: https://www.desmos.com/calculator/mtj2rdv9nm

And finally, here is your code updated to use the new formula:

```import * as THREE from "https://cdn.jsdelivr.net/npm/three@0.170/build/three.webgpu.js/+esm";
import { OrbitControls } from "https://cdn.jsdelivr.net/npm/three@0.170/examples/jsm/controls/OrbitControls.js/+esm";
import { wgslFn, varyingProperty, vec4, cameraProjectionMatrix, cameraViewMatrix, modelWorldMatrix, attribute, uniform } from "https://esm.sh/three/tsl";

const renderer = new THREE.WebGPURenderer({ antialias: true, logarithmicDepthBuffer: true });
renderer.outputColorSpace = THREE.SRGBColorSpace;
renderer.setPixelRatio( window.devicePixelRatio );
renderer.setSize( window.innerWidth, window.innerHeight );
renderer.setClearColor( 0x000000 );
renderer.setClearAlpha( 0 );
document.body.appendChild( renderer.domElement );

const aspect = window.innerWidth / window.innerHeight;

const camera = new THREE.PerspectiveCamera( 50.0, aspect, 1e-6, 1e12 );
const scene = new THREE.Scene();

scene.background = new THREE.Color( 0x00001f );
camera.position.set( 30, 40, 140 );
const controls = new OrbitControls( camera, renderer.domElement );

window.addEventListener( "resize", onWindowResize, false );
//--------------------------------------------------------------------------

await init();
await render();

async function init() {

  await renderer.init();
  
  const cameraNear = uniform( Math.max(1e-6, camera.near) );
  const cameraFar = uniform( camera.far );  
  
  const perspectiveW = varyingProperty("float", "perspectiveW");  //varying
  
  //params contain uniforms and attributes
  const vertexShaderParams = {
    projectionMatrix: cameraProjectionMatrix,
    cameraViewMatrix: cameraViewMatrix,
    modelWorldMatrix: modelWorldMatrix,  
    position: attribute("position")   
  }
  
  const vertexShader = wgslFn(`
    fn main_vertex(
      projectionMatrix: mat4x4<f32>,
      cameraViewMatrix: mat4x4<f32>,
      modelWorldMatrix: mat4x4<f32>,
      position: vec3<f32>
    ) -> vec4<f32> {
    
      var outPosition = projectionMatrix * cameraViewMatrix * modelWorldMatrix * vec4<f32>(position, 1);
      
      varyings.perspectiveW = outPosition.w;
      
      return outPosition;
    }
  `, [perspectiveW]); //add the varying
  
  //varying and uniform
  const depthShaderParams = {
    perspectiveW: perspectiveW,
    cameraNear: cameraNear,
    cameraFar: cameraFar
  }
  
  const depthShader = wgslFn(`
    fn main_depth(
      perspectiveW: f32,
      cameraNear: f32,
      cameraFar: f32
    ) -> vec4<f32> {
    
      var depth = log2( ( perspectiveW / cameraNear ) + 1 ) / log2( ( cameraFar / cameraNear ) + 1 );
      return vec4<f32>(depth);
    }
  `);
  
  const geometry = new THREE.BoxGeometry( 20, 60, 60 );
  const material = new THREE.MeshBasicNodeMaterial();
  material.vertexNode = vertexShader( vertexShaderParams );
  material.depthNode = depthShader( depthShaderParams );
  material.colorNode = vec4( 1, 0.25, 1, 1 ); //a fragment shader would also be possible
 
  const box = new THREE.Mesh( geometry, material );
  box.position.set( 0, 0, 0 );
  box.rotation.y = Math.PI / 2;
  const box2 = box.clone();
  box2.position.set( 0, 0, 100 );
  
  scene.add( box ); 
  scene.add( box2 );
  
  
  const selectableGeometry = new THREE.BoxGeometry(100, 100, 100);
  const selectableMaterial = new THREE.MeshBasicNodeMaterial({
    color: 0x12ffff,
    transparent: true,
    opacity: 0.4,
    side: THREE.DoubleSide
  });

  const lineMaterial = new THREE.LineBasicMaterial({
    color: 0xffffff,
    linewidth: 2,
  });

  const selectableMesh = new THREE.Mesh(selectableGeometry, selectableMaterial);
  selectableMesh.name = `Building_${1}_SelectableBox_Z${1}_Y${1}`;
  selectableMesh.position.set(0, 0, -20);

  const edges = new THREE.EdgesGeometry(selectableGeometry);
  const line = new THREE.LineSegments(edges, lineMaterial);

  selectableMesh.add(line);
  
  scene.add(selectableMesh);

 
  const selectableMesh2 = new THREE.Mesh(selectableGeometry, selectableMaterial);
  selectableMesh2.position.set(0, 0, -200);

  const edges2 = new THREE.EdgesGeometry(selectableGeometry);
  const line2 = new THREE.LineSegments(edges2, lineMaterial);
  selectableMesh2.add(line2);

  scene.add(selectableMesh2);  
}

//---------------------------------------------------------------
function render() {
  requestAnimationFrame( render );
  renderer.render(scene, camera);
}

 
function onWindowResize() {
  camera.aspect = window.innerWidth / window.innerHeight;
  camera.updateProjectionMatrix();
  renderer.setSize( window.innerWidth, window.innerHeight );
}
```

Spiri0 commented:
I was already looking at your change in detail. Thanks for the explanation. That helps me understand. Very interesting that logDepthBufFC is no longer neccessary. But I always found it strange that only the far value appeared in it. Your formula seems more complete to me.

The logarithmicDepthBuffer has been a topic in the forum a few times. I only recently created the codePen based on a user request in the forum.
I will keep the node variant and the shader variant in there, so users can see both.

Thank you for your effort. I work a lot with the logarithmicDepthBuffer.

Mugen87 commented on src/nodes/display/ViewportDepthNode.js:
```diff
@@ -116,6 +116,31 @@ export const viewZToPerspectiveDepth = ( viewZ, near, far ) => near.add( viewZ )
 // maps perspective depth in [ 0, 1 ] to viewZ
 export const perspectiveDepthToViewZ = ( depth, near, far ) => near.mul( far ).div( far.sub( near ).mul( depth ).sub( far ) );
 
+export const perspectiveDepthToLogarithmicDepth = ( perspectiveW, near, far ) => {
```
In post processing we have to compute the view position of a fragment in various effects. We do that with `getViewPosition()`:

https://github.com/mrdoob/three.js/blob/9e5528d2890de5432c6472bde6702abecc8e1cf7/src/nodes/utils/PostProcessingUtils.js#L15

This method does currently not work if a logarithmic Depth buffer is used. How would the reverse operation of `perspectiveDepthToLogarithmicDepth()` look like that computes a "normal" depth value based on the logarithmic depth?

PoseidonEnergy replied:
@Mugen87 Computing the perspective `w` value from a logarithmic depth value using the new formula would look like this. Note that this gives you the `w` value, as in `modelViewProjection.w`. To get the "actual" `z` value, just divide `z` by `w` (I know you know that already but I'm just saying it for posterity).

![image](https://github.com/user-attachments/assets/78ae6f02-ea51-4a52-afcc-11320f74491e)

The TSL node function based on the math shown above would look like this. Like mentioned before, this returns the perspective `w` value, and not an "actual" depth. To get the actual depth, just do `z / w`:

```
const logarithmicDepthToPerspectiveDepth = (logDepth, near, far) => {
  // NOTE: this returns the "w" component, not "z"
  const exponent = logDepth.mul(log(far.div(near).add(1)));
  const e = 2.718281828459045235360287471352;
  return float(e).pow(exponent).sub(1).mul(near);
};
```

**NOTE: I have not tested the above node function above...it looks correct though...needs another set of eyes.**

As an aside, should the functions `perspectiveDepthToLogarithmicDepth()` and `logarithmicDepthToPerspectiveDepth()` be renamed to `perspectiveWToLogarithmicDepth()` and `logarithmicDepthToPerspectiveW()` to make it more clear that the `w` component is being used?

Desmos graph: https://www.desmos.com/calculator/1e3ttyfe6q

![image](https://github.com/user-attachments/assets/506a12bf-bd99-430c-a2b2-48022418792f)

Mugen87 replied:
I've tried to replace the `sampleDepth()` function in `GTAONode` for quick testing but it does not work yet. Here is the code:
```js
const sampleDepth = ( uv ) => {

	const depth = this.depthNode.uv( uv ).x;

	if ( builder.renderer.logarithmicDepthBuffer === true ) {

		const mvp = positionWorld.mul( this.cameraProjectionMatrix );

		return mvp.z.div( logarithmicDepthToPerspectiveDepth( depth, this.cameraNear, this.cameraFar ) );

	} else {

		return depth;

	}

};
```
Do you see an obvious issue with the above code?

PoseidonEnergy replied:
I can say the following at the moment:

1. `depth` in `const depth = this.depthNode.uv( uv ).x;` should range between 0 and 1 before being passed to `logarithmicDepthToPerspectiveDepth( depth, this.cameraNear, this.cameraFar )`
2. Without looking at the rest of the code, `this.cameraNear` and `this.cameraFar` should be uniforms

In the meantime I will continue verifying that the `logarithmicDepthToPerspectiveDepth` is correct.

Mugen87 replied:
Both requirements were true when during my tests. I'll revisit the code tomorrow.

PoseidonEnergy replied:
Does the final `depth` value returned from `sampleDepth` need to be linear/orthographic? If so, there should be a conversion from viewZ value you get after the division by `w` to orthographic:

```
const viewZ = mvp.z.div( logarithmicDepthToPerspectiveDepth( depth, this.cameraNear, this.cameraFar ) );

return viewZToOrthographicDepth( viewZ, this.cameraNear, this.cameraFar );

I am still confirming that the logarithmicDepthToPerspectiveDepth function is correct.
```

Mugen87 replied:
> Does the final depth value returned from sampleDepth need to be linear/orthographic?

No, at least the AO uses the raw depth from the depth texture attachment.

Mugen87 replied:
> As an aside, should the functions perspectiveDepthToLogarithmicDepth() and logarithmicDepthToPerspectiveDepth() be renamed to perspectiveWToLogarithmicDepth() and logarithmicDepthToPerspectiveW() to make it more clear that the w component is being used?

Yes, I think otherwise the names are confusing. Let's use `perspectiveWToLogarithmicDepth()`.

Mugen87 replied:
I can't get it to work with the proposed formula. I've also tried a slightly different formula but with no success.
```js
export const logarithmicDepthToPerspectiveW = ( logarithmicW, near, far ) => {

	const logRatio = log2( far.div( near ).add( 1 ) );
	const perspectiveW = near.mul( pow( 2, logarithmicW.mul( logRatio ).sub( 1 ) ) );
	return perspectiveW;

};
```

PoseidonEnergy replied:
Perhaps your `mvp.z` equals (or is very close to) `mvp.w`? You may be returning `1` after dividing `z` by `w`.

Here is a fiddle where `mvp.z` equals `mvp.w`: https://jsfiddle.net/h7sd13pj

If that is the case, then try returning just the inverse:

`return float( 1 ).div( logarithmicDepthToPerspectiveDepth( depth, this.cameraNear, this.cameraFar ) );`

PoseidonEnergy replied:
Can you tell me how I can use `GTAONode` on my own to help you work on this? Maybe some little example or screenshot of what works when `logarithmicDepthBuffer` is `false` compared to when it is `true`?

Mugen87 replied:
There is an example for demonstrating the AO: https://threejs.org/examples/webgpu_postprocessing_ao

I suggest you test this locally by turning on logarithmic depth buffer in the demo. You will see the AO breaks. 

For testing I have added  `logarithmicDepthToPerspectiveW()` to `ViewportDepthNode` and then tried to use it in `GTAONode` (with no success so far).

PoseidonEnergy commented:
> I was already looking at your change in detail. Thanks for the explanation. That helps me understand. Very interesting that logDepthBufFC is no longer neccessary. But I always found it strange that only the far value appeared in it. Your formula seems more complete to me.
> 
> The logarithmicDepthBuffer has been a topic in the forum a few times. I only recently created the codePen based on a user request in the forum. I will keep the node variant and the shader variant in there, so users can see both.
> 
> Thank you for your effort. I work a lot with the logarithmicDepthBuffer.

@Spiri0 Hey there, just letting you know that the logarithmic depth buffer formula and related functions have undergone another revision which most definitely will affect your code in the near future. This revision is in PR #29870. Namely, the function `perspectiveDepthToLogarithmicDepth` has been renamed to `viewZToLogarithmicDepth`, and it expects a _negative_ `viewZ` value, and NOT a `w` component. In addition, the formula itself has been changed very slightly.

Here is the old calculation:

![image](https://github.com/user-attachments/assets/51e75f5b-ad64-4533-8aa9-5e8673385720)

And here is the new calculation:

![image](https://github.com/user-attachments/assets/658918e6-3862-424d-8a72-aac1cd22dbf1)

As mentioned before, this revision can be found in PR #29870.
